pkg/policy/selectorcache.go

```diff
@@ -695,132 +248,103 @@ func (l *labelIdentitySelector) fetchIdentityMappings() []identity.NumericIdenti
-	idsAsMap := make(map[identity.NumericIdentity]struct{}, len(identities))
-	for _, v := range identities {
-		if _, exists := idsAsMap[v]; exists {
-			identitiesToRelease = append(identitiesToRelease, v)
-		} else {
-			idsAsMap[v] = struct{}{}
-		}
+	fqdnSelect, ok := idSelector.source.(*fqdnSelector)
+	if !ok {
+		log.Error("UpdateFQDNSelector for non-FQDN selector!")
+		return UpdateResultUnchanged
 	}
 
+	// update wantLabels, then determine set of added and removed identities
+	fqdnSelect.setSelectorIPs(ips) // this updates wantLabels
+
 	// Note that 'added' and 'deleted' are guaranteed to be
 	// disjoint, as one of them is left as nil, or an identity
 	// being in 'identities' is a precondition for an
 	// identity to be appended to 'added', while the inverse is
 	// true for 'deleted'.
 	var added, deleted []identity.NumericIdentity
 
-	/* TODO - the FQDN side should expose what was changed (IPs added, and removed)
-	*  not all IPs corresponding to an FQDN - this will make this diff much
-	*  cheaper, but will require more plumbing on the FQDN side. for now, this
-	*  is good enough.
-	*
-	*  Case 1: identities did correspond to this FQDN, but no longer do. Reset
-	*  the map
-	 */
-	if len(identities) == 0 && len(fqdnSel.cachedSelections) != 0 {
-		// Need to update deleted to be all in cached selections
-		for k := range fqdnSel.cachedSelections {
-			deleted = append(deleted, k)
-			identitiesToRelease = append(identitiesToRelease, k)
-		}
-		fqdnSel.cachedSelections = make(map[identity.NumericIdentity]struct{})
-	} else if len(identities) != 0 && len(fqdnSel.cachedSelections) == 0 {
-		// Case 2: identities now correspond to this FQDN, but didn't before.
-		// We don't have to do any comparison of the maps to see what changed
-		// and what didn't.
-		added = identities
-		fqdnSel.cachedSelections = idsAsMap
-	} else {
-		// Case 3: Something changed resulting in some identities being added
-		// and / or removed. Figure out what these sets are (new identities
-		// added, or identities deleted).
-		for k := range fqdnSel.cachedSelections {
-			// If identity in cached selectors isn't in identities which were
-			// passed in, mark it as being deleted, and remove it from
-			// cachedSelectors.
-			if _, ok := idsAsMap[k]; !ok {
-				deleted = append(deleted, k)
-				delete(fqdnSel.cachedSelections, k)
-			}
-
-			// This function is passed a complete set of the new
-			// identities to associate with this selector, and each
-			// identity already has a reference count. Therefore,
-			// in order to balance references to the same
-			// identities, we should always remove references to
-			// identities that were preveiously selected by this
-			// selector.
-			identitiesToRelease = append(identitiesToRelease, k)
+	// Delete any non-matching entries from cachedSelections
+	for nid := range idSelector.cachedSelections {
+		identity := sc.idCache[nid]
+		if !idSelector.source.matches(identity) {
```

Comment:
I'm idly wondering whether the new path with more `matches()` calls may be more inefficient. Not sure whether this path is well set up for some benchmarking.

Reply:
Overall there is a bit more work within the SelectorCache, since we now scan all IDs on adding a FQDNSelector. However, we don't have to perform allocation, so that is probably a net win. In any case, FQDNSelectors and LabelIdentitySelectors now look basically the same, so that is a nice little cleanup.

This particular change doesn't actually add any more matches, it's just an interface cleanup.